File: newmx/_encoder.py
"""
NewMx Path 1 — encoder (internal).

Three-pass encoding:
1. Family pass — fixed-point iteration over conjunction-aware boundary regex
2. Orphan-suffix-after-glyph strip (rev4)
3. Structural pass — single combined alternation
4. Pre-glyph-space strip (cli5)

All four steps mirror path1_pipeline_b3_v005_rev4_FAST_cli5.py exactly.
"""

import re
from dataclasses import dataclass, field
from typing import Optional, List, Dict, Tuple

from ._constants import (
    INTENT_FAMILIES, BANNED_FAMILY_PHRASES, FAMILY_BOUNDARY_CONJUNCTIONS,
)


@dataclass
class CompiledPatterns:
    """Pre-compiled regex artifacts for fast encoding. Build once per codec."""
    family_pattern:       Optional[re.Pattern] = None
    family_dispatch:      Dict[str, Tuple[str, str]] = field(default_factory=dict)
    structural_pattern:   Optional[re.Pattern] = None
    structural_dispatch:  Dict[str, str] = field(default_factory=dict)
    orphan_suffix_pattern: Optional[re.Pattern] = None
    preglyph_space_pattern: Optional[re.Pattern] = None
    family_glyph_class:   str = ""
    all_glyph_class:      str = ""


def _build_family_glyph_class(phrase_glyph_map: dict) -> str:
    """Char-class string of distinct family glyphs (for boundary lookbehind).
    Skips ASCII Latin to keep \\b semantics sane in adjacent regexes."""
    glyphs = set()
    for info in phrase_glyph_map["phrase_to_family_and_glyph"].values():
        if info["family"] not in INTENT_FAMILIES:
            continue
        g = info["glyph"]
        if len(g) != 1:
            continue
        if "A" <= g <= "Z" or "a" <= g <= "z":
            continue
        glyphs.add(g)
    return "".join(re.escape(g) for g in sorted(glyphs))


def _build_all_glyph_class(phrase_glyph_map: dict) -> str:
    """Char-class string of all distinct glyphs (family + structural).
    Used by cli5 pre-glyph-space stripper."""
    glyphs = set()
    for info in phrase_glyph_map["phrase_to_family_and_glyph"].values():
        g = info["glyph"]
        if len(g) != 1:
            continue
        if "A" <= g <= "Z" or "a" <= g <= "z":
            continue
        glyphs.add(g)
    return "".join(re.escape(g) for g in sorted(glyphs))


def compile_patterns(phrase_glyph_map: dict) -> CompiledPatterns:
    """Compile all regexes needed for encoding. Run once per codec, reuse."""
    cp = CompiledPatterns()
    family_class = _build_family_glyph_class(phrase_glyph_map)
    all_class    = _build_all_glyph_class(phrase_glyph_map)
    cp.family_glyph_class = family_class
    cp.all_glyph_class    = all_class

    # ---- Family-boundary lookbehind (rev3) ----
    # Matches: line start, sentence-ender, any family glyph, or coordinating
    # conjunction + space. JS-style fixed-width lookbehinds for regex engine
    # compatibility (Python re supports variable-width but we keep parity).
    boundary_parts = ["^", "(?<=[.!?:;,\\n])"]
    if family_class:
        boundary_parts.append("(?<=[" + family_class + "])")
    for conj in FAMILY_BOUNDARY_CONJUNCTIONS:
        boundary_parts.append(r"(?<=\b" + conj + r"\s)")
    boundary_lookbehind = "(?:" + "|".join(boundary_parts) + ")"

    # ---- Sort entries longest-first (so 'tell me what you find' wins over
    # 'tell me') ----
    phrases = []
    for phrase, info in phrase_glyph_map["phrase_to_family_and_glyph"].items():
        if phrase in BANNED_FAMILY_PHRASES:
            continue
        phrases.append((phrase, info))
    phrases.sort(key=lambda x: (-len(x[0].split()), -len(x[0])))

    # ---- Partition into family vs structural ----
    family_entries = []
    struct_entries = []
    for phrase, info in phrases:
        if info["family"] in INTENT_FAMILIES:
            family_entries.append((phrase, info))
        else:
            struct_entries.append((phrase, info))

    # ---- Family alternation pattern + dispatch ----
    if family_entries:
        alt = "|".join(re.escape(p) for p, _ in family_entries)
        cp.family_pattern = re.compile(
            boundary_lookbehind + r"\s*(" + alt + r")\b",
            re.IGNORECASE,
        )
        for phrase, info in family_entries:
            cp.family_dispatch[phrase.lower()] = (info["family"], info["glyph"])

    # ---- Structural alternation pattern + dispatch ----
    if struct_entries:
        alt = "|".join(re.escape(p) for p, _ in struct_entries)
        cp.structural_pattern = re.compile(
            r"\b(" + alt + r")\b",
            re.IGNORECASE,
        )
        for phrase, info in struct_entries:
            cp.structural_dispatch[phrase.lower()] = info["glyph"]

    # ---- Orphan-suffix-after-glyph stripper (rev4) ----
    # Strips "<family-glyph> for me / to me / with me" — the glyph already
    # encodes the speaker-beneficiary relationship.
    if family_class:
        cp.orphan_suffix_pattern = re.compile(
            r"([" + family_class + r"])\s+(?:for|to|with)\s+me\b",
            re.IGNORECASE,
        )

    # ---- Pre-glyph-space stripper (cli5) ----
    # Strips whitespace immediately BEFORE any glyph (family or structural).
    # cl100k_base BPE merges " and" but not " 元", so each glyph boundary
    # paid an orphan-space token in cli4. This collapses that waste.
    if all_class:
        cp.preglyph_space_pattern = re.compile(
            r"\s+([" + all_class + r"])"
        )

    return cp


def encode_line(line: str, compiled: CompiledPatterns) -> Tuple[str, List[Dict]]:
    """Encode one line of pre-normalized text. Returns (encoded_text, replacements_list).

    Each replacement dict has keys: family, phrase, matched, glyph.
    """
    text = line
    replacements: List[Dict] = []

    # Family pass — fixed-point iteration (max 4 passes empirically sufficient,
    # range(4) is the safety bound from convergence audit).
    if compiled.family_pattern is not None and compiled.family_pattern.search(text):
        for _pass in range(4):
            before = text

            def family_repl(match: re.Match) -> str:
                matched = match.group(1)
                key = matched.lower()
                info = compiled.family_dispatch.get(key)
                if info is None:
                    return matched
                family, glyph = info
                replacements.append({
                    "family": family,
                    "phrase": key,
                    "matched": matched,
                    "glyph": glyph,
                })
                return glyph

            text = compiled.family_pattern.sub(family_repl, text)
            if text == before:
                break

    # Orphan-suffix-after-glyph strip (rev4)
    if compiled.orphan_suffix_pattern is not None:
        text = compiled.orphan_suffix_pattern.sub(r"\1", text)

    # Structural pass — single regex
    if compiled.structural_pattern is not None:
        def struct_repl(match: re.Match) -> str:
            matched = match.group(1)
            key = matched.lower()
            glyph = compiled.structural_dispatch.get(key)
            if glyph is None:
                return matched
            replacements.append({
                "family": "STRUCTURAL",
                "phrase": key,
                "matched": matched,
                "glyph": glyph,
            })
            return glyph

        text = compiled.structural_pattern.sub(struct_repl, text)

    # Pre-glyph-space strip (cli5)
    if compiled.preglyph_space_pattern is not None:
        text = compiled.preglyph_space_pattern.sub(r"\1", text)

    return text, replacements

File: android/app/src/main/java/com/cccstudios/newmx/codec/Encoder.kt
package com.cccstudios.newmx.codec

import java.util.regex.Pattern

/**
 * Pre-compiled regex artifacts for fast encoding. Build once per codec, reuse.
 *
 * Mirrors Python newmx/_encoder.py::CompiledPatterns.
 *
 * Patterns held here are stateless and thread-safe (java.util.regex.Pattern
 * is immutable; Matchers are not, but we create one per call).
 */
data class CompiledPatterns(
    val familyPattern: Pattern? = null,
    val familyDispatch: Map<String, FamilyEntry> = emptyMap(),
    val structuralPattern: Pattern? = null,
    val structuralDispatch: Map<String, String> = emptyMap(),
    val orphanSuffixPattern: Pattern? = null,
    val preglyphSpacePattern: Pattern? = null,
    val familyGlyphClass: String = "",
    val allGlyphClass: String = ""
) {
    data class FamilyEntry(val family: String, val glyph: String)
}

/**
 * Encoder — Kotlin port of newmx/_encoder.py.
 *
 * Logic mirror is exact:
 *   1. Family pass (fixed-point iteration, boundary-aware)
 *   2. Orphan-suffix-after-glyph strip (rev4)
 *   3. Structural pass
 *   4. Pre-glyph-space strip (cli5)
 *
 * IMPORTANT REGEX DIFFERENCES:
 *   - Python case-insensitive matching of non-ASCII works with re.IGNORECASE +
 *     re.UNICODE. Java equivalent: Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE.
 *   - Pattern.compile is happy with the same regex syntax as Python re for our
 *     use cases (alternation, lookbehind, char classes).
 *
 * NB: variable-width lookbehind is needed because of the boundary_lookbehind
 * structure ("|^|(?<=[.!?:;,\\n])|(?<=[GLYPHS])|(?<=\\band\\s)|(?<=\\bor\\s)..."
 * which has alternatives of different widths). Java's regex engine supports
 * this since Java 9 via the (?<=...) variable-width syntax with bounded
 * alternation. Verified: \b<conj>\s is bounded (max 5 chars: "then "), so OK.
 */
object Encoder {

    /**
     * Build the family glyph character class (used in boundary lookbehind).
     *
     * Returns a string containing all distinct family-layer glyph characters,
     * each regex-escaped. Same filtering as Python:
     *   - Family-layer entries only
     *   - Single-character glyphs only
     *   - Skip ASCII A-Za-z (would break \b semantics)
     */
    private fun buildFamilyGlyphClass(map: PhraseGlyphMap): String {
        val glyphs = sortedSetOf<String>()
        for (entry in map.entries) {
            if (entry.family !in CodecConstants.INTENT_FAMILIES) continue
            val g = entry.glyph
            if (g.length != 1) continue
            val c = g[0]
            if (c in 'A'..'Z' || c in 'a'..'z') continue
            glyphs.add(g)
        }
        return glyphs.joinToString("") { Pattern.quote(it).removeSurrounding("\\Q", "\\E").let { s -> regexEscapeChar(s) } }
    }

    /**
     * Build the "all glyph" character class (used by cli5 pre-glyph-space).
     *
     * Same as family class but includes structural glyphs too.
     */
    private fun buildAllGlyphClass(map: PhraseGlyphMap): String {
        val glyphs = sortedSetOf<String>()
        for (entry in map.entries) {
            val g = entry.glyph
            if (g.length != 1) continue
            val c = g[0]
            if (c in 'A'..'Z' || c in 'a'..'z') continue
            glyphs.add(g)
        }
        return glyphs.joinToString("") { regexEscapeChar(it) }
    }

    /**
     * Regex-escape a single character for inclusion in a character class.
     * Inside [...], only \, ], ^, - need escaping; everything else is literal.
     */
    private fun regexEscapeChar(s: String): String {
        if (s.length != 1) return Pattern.quote(s)
        return when (s[0]) {
            '\\', ']', '^', '-' -> "\\$s"
            else -> s
        }
    }

    /**
     * Regex-escape a multi-character phrase for use in an alternation.
     * Equivalent to Python's re.escape(): escapes anything that's a regex metachar.
     */
    private fun regexEscapePhrase(s: String): String {
        val sb = StringBuilder(s.length + 4)
        for (c in s) {
            when (c) {
                '\\', '.', '+', '*', '?', '(', ')', '|', '[', ']', '{', '}',
                '^', '$', '#', '&', '-', '~' -> sb.append('\\').append(c)
                else -> sb.append(c)
            }
        }
        return sb.toString()
    }

    /**
     * Compile all patterns. Run once per codec load (~50-100ms on first call,
     * then cached on the codec instance).
     */
    fun compilePatterns(map: PhraseGlyphMap): CompiledPatterns {
        val familyClass = buildFamilyGlyphClass(map)
        val allClass = buildAllGlyphClass(map)

        // ---- Family-boundary lookbehind ----
        // Matches: line start, sentence-ender, any family glyph, or
        // conjunction + whitespace.
        val boundaryParts = mutableListOf<String>()
        boundaryParts.add("^")
        boundaryParts.add("(?<=[.!?:;,\\n])")
        if (familyClass.isNotEmpty()) {
            boundaryParts.add("(?<=[$familyClass])")
        }
        for (conj in CodecConstants.FAMILY_BOUNDARY_CONJUNCTIONS) {
            // Variable-width lookbehind. Bounded (conj+space = max 5 chars),
            // accepted by Java regex engine.
            boundaryParts.add("(?<=\\b$conj\\s)")
        }
        val boundaryLookbehind = "(?:" + boundaryParts.joinToString("|") + ")"

        // ---- Sort entries longest-first ----
        // Critical: "tell me what you find" must match before "tell me".
        // Tiebreaker on char length keeps order deterministic.
        val entries = map.entries
            .filter { it.phrase !in CodecConstants.BANNED_FAMILY_PHRASES }
            .sortedWith(
                compareByDescending<PhraseGlyphMap.Entry> { it.phrase.split(' ').size }
                    .thenByDescending { it.phrase.length }
            )

        // ---- Partition family vs structural ----
        val familyEntries = entries.filter { it.family in CodecConstants.INTENT_FAMILIES }
        val structEntries = entries.filter { it.family !in CodecConstants.INTENT_FAMILIES }

        // ---- Family alternation ----
        var familyPattern: Pattern? = null
        val familyDispatch = mutableMapOf<String, CompiledPatterns.FamilyEntry>()
        if (familyEntries.isNotEmpty()) {
            val alt = familyEntries.joinToString("|") { regexEscapePhrase(it.phrase) }
            val patternStr = boundaryLookbehind + "\\s*(" + alt + ")\\b"
            familyPattern = Pattern.compile(
                patternStr,
                Pattern.CASE_INSENSITIVE or Pattern.UNICODE_CASE
            )
            for (e in familyEntries) {
                familyDispatch[e.phrase.lowercase()] = CompiledPatterns.FamilyEntry(e.family, e.glyph)
            }
        }

        // ---- Structural alternation ----
        var structuralPattern: Pattern? = null
        val structuralDispatch = mutableMapOf<String, String>()
        if (structEntries.isNotEmpty()) {
            val alt = structEntries.joinToString("|") { regexEscapePhrase(it.phrase) }
            structuralPattern = Pattern.compile(
                "\\b(" + alt + ")\\b",
                Pattern.CASE_INSENSITIVE or Pattern.UNICODE_CASE
            )
            for (e in structEntries) {
                structuralDispatch[e.phrase.lowercase()] = e.glyph
            }
        }

        // ---- Orphan-suffix-after-glyph (rev4) ----
        var orphanSuffix: Pattern? = null
        if (familyClass.isNotEmpty()) {
            orphanSuffix = Pattern.compile(
                "([$familyClass])\\s+(?:for|to|with)\\s+me\\b",
                Pattern.CASE_INSENSITIVE or Pattern.UNICODE_CASE
            )
        }

        // ---- Pre-glyph-space (cli5) ----
        var preglyphSpace: Pattern? = null
        if (allClass.isNotEmpty()) {
            preglyphSpace = Pattern.compile("\\s+([$allClass])")
        }

        return CompiledPatterns(
            familyPattern = familyPattern,
            familyDispatch = familyDispatch,
            structuralPattern = structuralPattern,
            structuralDispatch = structuralDispatch,
            orphanSuffixPattern = orphanSuffix,
            preglyphSpacePattern = preglyphSpace,
            familyGlyphClass = familyClass,
            allGlyphClass = allClass
        )
    }

    /**
     * Replacement record (mirror of Python's dict).
     */
    data class Replacement(
        val family: String,
        val phrase: String,
        val matched: String,
        val glyph: String
    )

    data class EncodeResult(
        val encoded: String,
        val replacements: List<Replacement>
    )

    /**
     * Encode one pre-normalized line.
     *
     * Returns (encoded_text, replacements_list).
     *
     * Family pass iterates to a fixed point (max 4 passes, same as Python).
     */
    fun encodeLine(line: String, cp: CompiledPatterns): EncodeResult {
        var text = line
        val reps = mutableListOf<Replacement>()

        // Family pass — fixed-point iteration
        if (cp.familyPattern != null && cp.familyPattern.matcher(text).find()) {
            for (pass in 0 until 4) {
                val before = text
                text = replaceAll(cp.familyPattern, text) { match ->
                    val matched = match.group(1) ?: return@replaceAll match.group()!!
                    val entry = cp.familyDispatch[matched.lowercase()] ?: return@replaceAll matched
                    reps.add(Replacement(entry.family, matched.lowercase(), matched, entry.glyph))
                    entry.glyph
                }
                if (text == before) break
            }
        }

        // Orphan-suffix-after-glyph (rev4)
        if (cp.orphanSuffixPattern != null) {
            text = cp.orphanSuffixPattern.matcher(text).replaceAll("$1")
        }

        // Structural pass
        if (cp.structuralPattern != null) {
            text = replaceAll(cp.structuralPattern, text) { match ->
                val matched = match.group(1) ?: return@replaceAll match.group()!!
                val glyph = cp.structuralDispatch[matched.lowercase()] ?: return@replaceAll matched
                reps.add(Replacement("STRUCTURAL", matched.lowercase(), matched, glyph))
                glyph
            }
        }

        // Pre-glyph-space (cli5)
        if (cp.preglyphSpacePattern != null) {
            text = cp.preglyphSpacePattern.matcher(text).replaceAll("$1")
        }

        return EncodeResult(text, reps)
    }

    /**
     * Helper: scan and replace with a function (mirror of Python re.sub with callback).
     *
     * Necessary because Java's Matcher.appendReplacement does $-substitution that
     * would mangle replacement strings containing literal $ or \.
     */
    private inline fun replaceAll(
        pattern: Pattern,
        text: String,
        replacement: (java.util.regex.MatchResult) -> String
    ): String {
        val m = pattern.matcher(text)
        val sb = StringBuilder(text.length)
        var last = 0
        while (m.find()) {
            sb.append(text, last, m.start())
            sb.append(replacement(m.toMatchResult()))
            last = m.end()
        }
        sb.append(text, last, text.length)
        return sb.toString()
    }
}
